### PR TITLE
Replace VarlenType with VarlenBits; add ABI version translation

### PR DESCRIPTION
**Pull request by xinyazhang**
# Overview

## Background

The `TH` (`Total_Q_Tokens`, `H_q`) memory layout of Log-Sum-Exp (LSE) tensor
from Variable Length Layout (Varlen) inputs was not supported in previous
AOTriton releases, which always use LSE layout generalized from
non-varlen inputs (BH, S) -> (H, T), by treating the batch axis `B` as `1`, and
sequence axis `S` as stacked total sequences `T`.

Unfortunately, the TH memory was used by TransformerEngine and expected by
PyTorch's `torch.nn.attention.varlen` API, and it's a challenge to extend the
existing VarlenType for various demands.

To address all the challenges, this PR replaces the `VarlenType` enum with
`VarlenBits`, a bit-field struct that decomposes varlen into three independent
per-side axes:

* Is the token axis stacked?
* How a length is given?
* Where a sequence starts?

plus a two-bit `LSE_LAYOUT` field, and ports the five Triton kernel
bodies (`attn_fwd`, `bwd_preprocess`, `bwd_kernel_dk_dv`, `bwd_kernel_dq`,
`bwd_kernel_fuse`), the C++ shims, the ATI/tuning descriptions, and the pybind
layer onto it.

In addition, this PR also formally introduces ABI backward compatibility
support, by translating parameter struct from older kVersion to current kVersion.
Thus, a binary built against the 0.13b headers keeps running when this library
is dropped in beneath it. **CAVEAT**, this feature does not provide source
compatibility, to build against newer AOTriton library, users are still
expected to use new API.

## Major Changes

* **BREAKING** [api] `VarlenType` is removed from `include/aotriton/flash.h`
  entirely (enum and `varlen_type` field). Replaced by `VarlenBits` — a
  struct holding one `VarlenMode` bit-field (`stacked`/`length`/`position`)
  per side, named `qmode`/`kmode`, plus a `lse_layout` field.
  + Configurations the enum could not spell: mixed per-side addressing
    (packed Q against dense K, `0x000B`), `seqused_k` on a packed KV cache
    (individual length + array position on K only), and TE's `(T, H)`
    logsumexp layout (`VarlenLseLayout::TH`, default remains `HT`).
  + Mixed-mode test coverage: compact Q against dense K (`0x000B`), and
    compact Q against `seqused_k`-style K on both a packed (`0x150B`) and a
    BHSD (`0x040B`) cache — covering `torch.nn.attention.varlen`'s
    `seqused_k`.
     - Other mixed-mode is possible, but untested.
* [shim] Add `modules/flash/csrc/varlen.h`, internal to the shims, to turn a
  `VarlenBits` into the `int32` word the kernel decodes and back.
  + `varlen_to_wire`/`varlen_from_wire` to translate b/w GPU format
  + Also keeps the retired `VarlenType` enum and `varlen_bits_of()`, which
    reconstructs the four legacy rows for the ABI translation path below.
* [kernel] All `flash` kernels now take `varlen_bits` instead
  + `varlen_bits.py` is added as the common decoder/addresser.
  + `Num_seqlens` is removed from Triton kernel.
* [kernel] All `flash` kernels now use `seqinfo_q/k0` and `seqinfo_q/k1`.
  + The old name `cu_seqlens_q/k`, `seq_strides_q/k` cannot cover all cases of `varlen_bits`.
* [kernel] `bwd_preprocess_varlen` is fused into `bwd_preprocess`
  + Consequently the metro kernel is updated
    - metro kernel transpiler test now uses `synthetic_metro_if_else` instead,
      since the branch of the real metro kernel has gone.
  + Kernel numbers within `attn_options` are re-computed
* [abi] Stable ABI is implemented:
  + Overall effect: A binary compiled against AOTriton 0.13b still runs.
  + `attn_fwd`/`attn_bwd` translate an older param object passed-in to current version.
  + The version is specified with `params_version`
  + Add new internal header `modules/flash/csrc/params_abi_compat.h` to:
    - Keep old layouts as `versioned_attn_{fwd,bwd}_params<N>`
    - Implement translation function within each class
    - Add macro to perform the translation automatically
  + **Coding Style** New User APIs should follow this practice to implement stable ABI

## Minor Changes

* [kernel] Fix a pre-existing LSE early-exit bug in `attn_fwd`: the causal
  early-exit path masked the `+inf` LSE write with `offs_m < Max_seqlen_q`,
  but under a stacked (compact varlen) layout the rows past `seqlen_q` belong
  to the *next* sequence, so it wrote `+inf` into another sequence's LSE. Now
  masked with `offs_m < seqlen_q`. Invisible until the varlen gate became
  bitwise.
* [shim] `attn_fwd.cc`/`attn_bwd.cc` validate `varlen_bits` via a per-side
  `VarlenAddressing`, returning `hipErrorInvalidValue` instead of launching a
  wrong-shaped grid:
  + Rejects out-of-range `length`/`position`, `REUSE` without `CUMULATIVE`,
    a missing required seqinfo array, and an invalid `lse_layout`.
  + Requires `Max_seqlen_{q,k}` whenever a mode reads it, catching a
    forgotten zero that previously launched zero workgroups and returned
    `hipSuccess` silently.
  + Checks seqinfo/tensor extents cover at least `N` sequences.
  + Independently recomputes the persistent-loop predicate
    (`Varlen_bits & 0xFFFF`) and the sequence count (`seq_count()` vs.
    `independent_seq_count()`) — both must agree exactly with the kernel's
    own check.
* [aiter] `aiter_fwd.cc`/`aiter_bwd.cc` reject any varlen inputs, for now.
* [test] Add JIT-level tests at `modules/flash/kernel/test_varlen_bits.py`
  to test the new `varlen_bits` Kernel in Triton JIT, comparing bitwise
  against N separate dense calls rather than a tolerance-based reference.
  - Consequently, rewrite `varlen_attn_torch_function.py`
* [test] `modules/flash/tests/test_varlen.py`/`test_op_bwd` thread
  `lse_layout` through the AOT API and add `test_logsumexp_layout`, pinning
  values against a closed form so a swapped Q/K-side index is visible (an
  all-equal buffer, used previously, compares equal to its own transpose).
* [test] Standalone `test_varlen_translation.cc` to test the struct
  translations.
* [test] Fix `test_logsumexp_scaling` never running as part of
  `test_backward.py`
  + Rename to `core_test_logsumexp_scaling` and call it within `test_backward.py`

# Tests Performed

This PR is tested with

```
bash .ci/build-test.sh gfx942 triton-3.7.0+gitdb82b800.aotriton0.14-cp313-cp313-linux_x86_64.whl
bash .ci/run-test.sh 0 0 split
bash .ci/run-test.sh 0 0 fused
```
within a Debian 13 container + ROCM 7.14.0 VENV.

### CODE REVIEW

File: include/aotriton/flash.h [hunk 1]
```diff
@@ -42,20 +42,24 @@ struct AOTRITON_API attn_options {
...
     attn_fwd = 0,
     debug_simulate_encoded_softmax = 1,
 
     // Backward pass kernels (from bwd_preprocess, bwd_kernel_*, etc.)
+    // bwd_preprocess_varlen was merged into bwd_preprocess (one kernel under
+    // varlen_bits == 0), so the dk_dv/dq/fuse slots shifted down by one. Safe
+    // because every consumer resolves this enum BY NAME: the shim codegen
+    // builds KERNEL_SLOT_INDEX from the kernel name, and modules/flash/tune/
+    // spells `c.KernelSlot.<name>`.
     bwd_preprocess = 2,
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply include/aotriton/flash.h:57]: Tuning build are not supposed to be compatible (even with in a dev cycle)

File: include/aotriton/flash.h [hunk 2]
```diff
@@ -151,25 +220,26 @@ struct AOTRITON_API attn_bwd_params {
...
   T0        philox_offset1;
   uint64_t  philox_offset2;
   int8_t    causal_type;
-  int8_t    varlen_type = 0;
+  // int8_t varlen_type;            // Superseded by varlen_bits
   int32_t   window_left;
   int32_t   window_right;
-  mutable LT4       DQ_ACC;          // fp32 accumulator of dq
+  mutable LT4       DQ_ACC;         // fp32 accumulator of dq
+  VarlenBits varlen_bits = {};
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply include/aotriton/flash.h:246]: The object will be translated at the funciton

File: modules/flash/kernel/test_varlen_bits.py [hunk 3]
```diff
@@ -0,0 +1,697 @@
+#!/usr/bin/env python
+# Copyright © 2025 Advanced Micro Devices, Inc.
+# SPDX-License-Identifier: MIT
+
+"""JIT-level gates for VarlenBits.
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply modules/flash/kernel/test_varlen_bits.py:5]: This is JIT tests, and should never go into AOT's tests/

File: modules/flash/tests/varlen_attn_torch_function.py [hunk 4]
```diff
@@ -194,13 +236,15 @@ def backward(ctx, do, _, __):
...
                               b, sm_scale, o, do, dq, dk, dv, db, dq_acc, L, delta,
-                              dropout_p, philox_seed, philox_offset, 0, causal, ctx.varlen_type, extargs);
+                              dropout_p, philox_seed, philox_offset, 0, causal, ctx.varlen_type,
+                              ctx.lse_layout, extargs);
         if PROBE_UNSUPPORTED and ret == hipError_t.hipErrorPeerAccessUnsupported:
             raise NotImplementedError()
         assert ret == hipError_t.hipSuccess, ret
-        return dq, dk, dv, None, None, None, None, None, None, None, None
+        # One None per non-tensor forward input, now including lse_layout.
+        return dq, dk, dv, None, None, None, None, None, None, None, None, None
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply modules/flash/tests/varlen_attn_torch_function.py]: lse_layout and attn_extra_args are not tensors
